### ISSUE TITLE
Requesting a **better** torch.auto_grad.detect_anomaly() for NaN detection

### ISSUE THREAD

**Issue by profPlum** [pytorch/pytorch#160016]
### 🚀 The feature, motivation and pitch

Hi, IMHO: There are several problems `with torch.auto_grad.detect_anomaly():` for NaN detection:
1. It raises errors inside pytorch internals which can't be debugged by pdb, it essentially just provides a traceback. 
Pytorch was originally created AFAIK to be more pythonic and easier to debug due to eager execution right? So why do this?
2. It's traceback points to the first place in the BACKWARDS pass where a grad function produces NaNs, which can easily be much later than the actual source of the NaNs (since backwards pass is in reverse), making the traceback it provides useless in many situations.
3. **It is possible for NaN loss values to pass silently through with** `torch.auto_grad.detect_anomaly():` which is highly counter-intuitive

Code example for \#2:
```
with torch.autograd.detect_anomaly():
    x = torch.tensor([1.0], requires_grad=True)
    y = torch.sqrt(x - 2) # root of the problem, produces first NaN
    y = torch.sqrt(y) # but detect_anomaly traceback points to this line
    y.backward()
```

Code example for \#3:
```
inp = torch.tensor([1.0], requires_grad=True)
with torch.autograd.detect_anomaly():
    loss = torch.log(-inp) # NaN
    loss.backward() # detect_anomaly raises **no error** despite NaN loss!
```

What I'm proposing is:
1. If possible raise a RuntimeError on the _forward_ operation that **would** create a NaN on the backwards pass even if e.g. it creates no NaN values during the forward pass itself. This way we can debug it like a regular python program.
2. If any NaN anywhere would to be computed within the scope torch.autograd.detect_anomaly() instead pytorch raises a RuntimeError on the spot, and if you set the config value globally then it should be recoverable with `try`/`except`.
3. \#3 honestly looks like a bug but if it is not then it is pathological intended behavior and shouldn't be allowed.

P.S. these code examples were tested on torch=='2.4.0.post202' and python==3.11.0

### Alternatives

Alternative to \#1 proposal is providing a lot more information in the traceback such as the values inside the derivative that caused the NaN. 
E.g. if f(x)=log(x) and log(1e-40).isfinite() but df/dx = 1/x = 1/1e-40='inf' then provide the information necessary to understand what is happening: like what the derivative looks like (1/x) and the value of x along with the traceback.

But really the pdb-compatible option is ideal, and consistent with the original motivations of Pytorch AFAIK.

**Comment by ezyang** [comment]
There are also funny things you can do with our compiler stack, like trace out the backwards into an FX graph and then pdb that, but it will still be auto-generated code that you will have to learn to interpret.

**Comment by profPlum** [comment]
@ezyang 
> This isn't possible. At forward time, we don't know what tangents will participate in the computation, and it might be the interaction of the tangent with the saved tensors that causes a NaN.
If I have something like y=sqrt(x) with requires_grad=True and you can see the forward value is x=0 this is an example where it is possible to detect during the forward the NaN which will happen in the backwards. dsqrt(x)/dx(0)=1/(2*sqrt(0))=NaN

> If you want to report at ANY time a NaN occurs, you can use something like https://github.com/albanD/subclass_zoo/blob/main/nan_detect.py
That was literally the original motivation of torch.auto_grad.detect_anomaly() [see the original issue](https://github.com/pytorch/pytorch/issues/1274#issue-222186988) we should finish the feature to include the functionality of the numpy feature that it was based off of rather than resorting to independent repos.

> because the backwards graph doesn't correspond to real Python code and so there really isn't a backtrace we can give; the forward that generated is the best we can do
I don't know exactly what you mean here? But you guys **are already** providing a backtrace. I'm requesting (among other things) that the backtrace is actually accurate in indicating the line which originally creates the NaNs.

On top of all these issues it doesn't even consistently raise an error when a NaN is created originally and only in the backwards pass! (see # 4 in my edited original post) This honestly might be a bug...

**Comment by albanD** [comment]
1. As mentioned above, this is unfortunately because the backward pass happens in C++. So there is no way for us to drop you into a nice debug environment there.
2. I guess that is a side effect of the fact that it only detects things during the backward. As it will flag the first NaN in any gradient in the backward. But indeed won't catch anything in the forward. In your exampe, it does point to the first NaN gradient, but that is indeed confusing when considering the forward NaNs.
3. Nothing is reported because there are no NaN gradients ever (even though the loss itself is NaN, the gradients is 1.). So this is expected (under the assumption that we only detect during the backward and not forward)
4. It is not reported because the gradient is `-inf` which is not `NaN`. And only the optimizer step makes this inf into a NaN.


But taking a step back, I think your ask is not torch.autograd.detect_anomaly() at all but an extended version of https://github.com/albanD/subclass_zoo/blob/main/nan_detect.py that would detect any gradient within the scope.

**Comment by profPlum** [comment]
@albanD 
> But taking a step back, I think your ask is not torch.autograd.detect_anomaly() at all but an extended version of https://github.com/albanD/subclass_zoo/blob/main/nan_detect.py that would detect any gradient within the scope?

This [was the original motivation](https://github.com/pytorch/pytorch/issues/1274#issue-222186988) of `torch.autograd.detect_anomaly()`: to raise on **any** NaN (like [numpy.seterr](https://docs.scipy.org/doc/numpy/reference/generated/numpy.seterr.html)).

So yes I'm requesting something like that be integrated into torch.autograd.detect_anomaly() **doing so would solve 2, 3 & even a little bit of 1. _But also requesting that it have a toggleable option to raise on infinity_ (for 4).**

P.S.

> 2\. I guess that is a side effect of the fact that it only detects things during the backward. As it will flag the first NaN in any gradient in the backward. But indeed won't catch anything in the forward. In your exampe, it does point to the first NaN gradient, but that is indeed confusing when considering the forward NaNs.

I've actually encountered a disheartening example where it will instead indicate a line that happens **earlier than** the actual source of the NaNs. I've been unable to get a minimal reproducible example yet though...

**Comment by profPlum** [comment]
@soulitzer 
Please also remember to add a toggleable option to raise on infinity (off by default), because #4 is a nasty example if you just get the NaN error inside the optimizer you don't know where it actually came from.
But otherwise that sounds good to me!
